File: server/src/main/kotlin/dev/kdrag0n/quicklock/server/Pairing.kt
package dev.kdrag0n.quicklock.server

import com.google.zxing.BarcodeFormat
import com.google.zxing.client.j2se.MatrixToImageWriter
import com.google.zxing.qrcode.QRCodeWriter
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.plugins.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okio.ByteString.Companion.toByteString
import java.security.SecureRandom

@Serializable
data class PairingChallenge(
    val id: String,
    val timestamp: Long,
    val isInitial: Boolean,
) {
    fun validate() {
        require(System.currentTimeMillis() - timestamp <= Config.TIME_GRACE_PERIOD)
    }
}

val pairingChallenges = HashMap<String, PairingChallenge>()
val finishPayloads = HashMap<String, String>()

@Serializable
private data class InitialPairQr(
    val secret: String,
)

@Serializable
private data class PairFinishPayload(
    val challengeId: String,
    val publicKey: String,
    val delegationKey: String,
    val blsPublicKeys: List<String>?,
    val mainAttestationChain: List<String>,
    val delegationAttestationChain: List<String>,
)

@Serializable
data class InitialPairFinishRequest(
    val finishPayload: String,
    val mac: String,
)

@Serializable
data class Delegation(
    val finishPayload: String,
    val expiresAt: Long,
    val allowedEntities: List<String>?,
)

@Serializable
private data class SignedDelegation(
    val device: String,
    val delegation: String,
    val blsSignature: String,
    val ecSignature: String,
)

fun generateSecret(): String {
    val secret = ByteArray(32)
    SecureRandom.getInstanceStrong().nextBytes(secret)
    return secret.toBase64()
}

@Volatile
var initialPairingSecret: String? = null

fun Application.pairingModule() = routing {
    fun finishPair(
        req: PairFinishPayload,
        delegatedBy: String?,
        expiresAt: Long = Long.MAX_VALUE,
        allowedEntities: List<String>? = null,
    ) {
        val challenge = pairingChallenges[req.challengeId]!!

        try {
            require(challenge.isInitial == (delegatedBy == null))

            // Verify timestamp
            challenge.validate()

            // Verify main attestation and certificate chain
            Attestation.verifyChain(req.mainAttestationChain, challenge.id)

            // Verify delegation attestation and certificate chain
            Attestation.verifyChain(req.delegationAttestationChain, challenge.id, isDelegation = true)

            Storage.addDevice(PairedDevice(
                publicKey = req.publicKey,
                delegationKey = req.delegationKey,
                // Params
                expiresAt = expiresAt,
                delegatedBy = delegatedBy,
                allowedEntities = allowedEntities,
            ))
        } finally {
            // Drop challenge
            pairingChallenges -= challenge.id
            finishPayloads -= challenge.id
        }
    }

    /*
     * Initial
     */
    post("/api/pair/initial/start") {
        // Only for initial setup
        require(!Storage.hasPairedDevices())
        require(initialPairingSecret == null)

        // Generate secret
        val secret = generateSecret()
        initialPairingSecret = secret

        // Open QR page locally (on server)
        withContext(Dispatchers.IO) {
            // URL-encode secret
            val url = URLBuilder().run {
                protocol = URLProtocol.HTTP
                host = "localhost"
                port = 3002
                path("api", "pair", "initial", "start", "qr")
                parameters.append("secret", secret)
                buildString()
            }

            println("secret = $secret")
            println("url = xdg-open '$url'")
            Runtime.getRuntime().exec("xdg-open '$url'")
        }

        call.respond(EmptyObject)
    }

    get("/api/pair/initial/start/qr") {
        require(call.request.origin.remoteHost == "localhost")

        val secret = initialPairingSecret!!
        require(call.request.queryParameters["secret"] == initialPairingSecret)

        val data = Json.encodeToString(InitialPairQr(
            secret = secret,
        ))
        val writer = QRCodeWriter()
        val matrix = writer.encode(data, BarcodeFormat.QR_CODE, 512, 512)
        call.respondOutputStream(ContentType.defaultForFileExtension("png")) {
            MatrixToImageWriter.writeToStream(matrix, "png", this)
        }
    }

    post("/api/pair/initial/finish") {
        val req = call.receive<InitialPairFinishRequest>()

        // Verify HMAC
        val secret = initialPairingSecret!!
        val expectedMac = req.finishPayload.serializeToByteArray().toByteString()
            .hmacSha256(secret.decodeBase64().toByteString())
            .toByteArray()
        require(expectedMac cryptoEq req.mac.decodeBase64())
        initialPairingSecret = null

        val payload = Json.decodeFromString<PairFinishPayload>(req.finishPayload)
        finishPair(payload, delegatedBy = null)

        call.respond(EmptyObject)
    }

    /*
     * Delegation
     */
    get("/api/pair/delegated/{challengeId}/finish_payload") {
        val id = call.parameters["challengeId"]!!

        val payload = finishPayloads[id]
        if (payload == null) {
            call.respond(HttpStatusCode.NotFound)
        } else {
            call.respondText(payload)
        }
    }

    post("/api/pair/delegated/{challengeId}/finish_payload") {
        val id = call.parameters["challengeId"]!!
        require(id in pairingChallenges)
        require(id !in finishPayloads)

        // Raw string, not decoded JSON, to avoid potential formatting differences
        // between Moshi and KotlinX serialization
        val payload = call.receiveChannel().readRemaining().readText()
        finishPayloads[id] = payload
        call.respond(EmptyObject)
    }

    post("/api/pair/delegated/{challengeId}/finish") {
        val req = call.receive<SignedDelegation>()
        val (delegatedBy, delegationData, signature) = req
        val device = Storage.getDeviceByKey(delegatedBy)
        Crypto.verifySignature(delegationData, device.delegationKey, signature)

        val (confirmedPayload, expiresAt, allowedEntities) = Json.decodeFromString<Delegation>(delegationData)

        // Prevent delegator from changing request
        val id = call.parameters["challengeId"]!!
        val origPayload = finishPayloads[id]!!
        require(origPayload == confirmedPayload)

        val payload = Json.decodeFromString<PairFinishPayload>(confirmedPayload)

        finishPair(
            payload,
            delegatedBy = delegatedBy,
            expiresAt = expiresAt,
            allowedEntities = allowedEntities,
        )

        call.respond(EmptyObject)
    }

    /*
     * Common
     */
    post("/api/pair/get_challenge") {
        // Generate a new challenge
        val challengeId = generateSecret()
        val challenge = PairingChallenge(
            id = challengeId,
            timestamp = System.currentTimeMillis(),
            isInitial = !Storage.hasPairedDevices(),
        )
        pairingChallenges[challengeId] = challenge

        call.respond(challenge)
    }
}

infix fun ByteArray.cryptoEq(other: ByteArray): Boolean {
    var n = 0
    for (i in indices) {
        n = n or (this[i].toInt() xor other[i].toInt())
    }
    return n == 0
}

File: server-rs/src/lock/pairing.rs
use std::net::SocketAddr;

use anyhow::anyhow;
use axum::response::IntoResponse;
use axum::{Json, Router};
use axum::extract::{Path, ConnectInfo};
use axum::routing::{get, post};
use dashmap::DashMap;
use lazy_static::lazy_static;
use parking_lot::Mutex;
use tracing::log::{debug, info};
use crate::crypto::hash;
use crate::envelope::SignedRequestEnvelope;
use crate::lock::model::{PairFinishPayload, PairingChallenge, InitialPairQr, InitialPairFinishRequest, Delegation};
use qrcode::QrCode;
use qrcode::render::unicode::Dense1x2;
use ring::hmac;
use crate::checks::require;
use super::request::EnvelopeOpen;
use super::store::{PairedDevice, STORE};
use crate::error::{AppResult, Error, HttpError};
use crate::time::now;
use super::attestation::verify_chain;
use super::CONFIG;
use super::crypto::generate_secret;

lazy_static! {
    static ref PAIRING_CHALLENGES: DashMap<String, PairingChallenge> = DashMap::new();
    static ref FINISH_PAYLOADS: DashMap<String, String> = DashMap::new();
    static ref INITIAL_PAIRING_SECRET: Mutex<Option<String>> = Mutex::new(None);
}

trait Validate {
    fn validate(&self) -> AppResult<()>;
}
impl Validate for PairingChallenge {
    fn validate(&self) -> AppResult<()> {
        require(now() - self.timestamp <= CONFIG.time_grace_period)?;
        Ok(())
    }
}

fn finish_pair(
    req: &PairFinishPayload,
    delegated_by: Option<String>,
    expires_at: u64,
    allowed_entities: Option<Vec<String>>,
) -> Result<(), Error> {
    let challenge = PAIRING_CHALLENGES.remove(&req.challenge_id)
        .map(|(_, challenge)| challenge)
        .ok_or(HttpError::NotFound)?;
    // Drop challenge data
    FINISH_PAYLOADS.remove(&challenge.id);

    require(challenge.is_initial == delegated_by.is_none())?;
    require(req.enc_key.len() == 32)?;

    // Verify timestamp
    challenge.validate()?;

    // Verify main attestation and certificate chain
    verify_chain(&req.main_attestation_chain, &challenge.id, false)?;

    // Verify delegation attestation and certificate chain
    verify_chain(&req.delegation_attestation_chain, &challenge.id, true)?;

    // Max expiry = delegator's expiry
    let expires_at = if let Some(ref delegator) = delegated_by {
        let delegator = STORE.get_device(delegator)
            .ok_or_else(|| anyhow!("Missing device"))?;
        expires_at.min(delegator.expires_at)
    } else {
        expires_at
    };

    // Only allow entities that delegator has access to
    let allowed_entities = match delegated_by {
        Some(ref delegator) => allowed_entities
            // Filter allowed list if given
            .map(|entities| {
                entities.into_iter()
                    .filter(|e| STORE.get_device_for_entity(delegator, e).is_some())
                    .collect()
            })
            // Otherwise limit to delegator's allowed list
            .or_else(|| STORE.get_device(delegator)
                .and_then(|d| d.allowed_entities)),
        None => allowed_entities,
    };

    // Enroll
    let pk_bytes = base64::decode(&req.public_key)?;
    STORE.add_device(PairedDevice {
        id: base64::encode(&hash(&pk_bytes)),
        public_key: req.public_key.clone(),
        delegation_key: req.delegation_key.clone(),
        enc_key: req.enc_key.clone(),
        audit_public_key: req.audit_public_key.clone(),
        // Params
        expires_at,
        delegated_by,
        allowed_entities,
    });
    Ok(())
}

/*
 * Initial
 */
async fn start_initial() -> AppResult<impl IntoResponse> {
    // Only for initial setup
    require(!STORE.has_paired_devices())?;
    require(INITIAL_PAIRING_SECRET.lock().is_none())?;

    // Generate secret
    let secret = generate_secret();
    debug!("secret = {}", secret);
    INITIAL_PAIRING_SECRET.lock().replace(secret.clone());

    // Print QR code
    let qr_data = serde_json::to_string(&InitialPairQr {
        secret,
    })?;
    let qr = QrCode::new(qr_data)?;
    let image = qr.render::<Dense1x2>().build();
    info!("{}", image);

    Ok(Json(()))
}

async fn finish_initial(req: Json<InitialPairFinishRequest>) -> AppResult<impl IntoResponse> {
    debug!("Finish initial pair: {:?}", req);

    // Verify HMAC
    let secret = INITIAL_PAIRING_SECRET.lock().take().unwrap();
    let key_bytes = base64::decode(&secret)?;
    let key = hmac::Key::new(hmac::HMAC_SHA256, &key_bytes);
    hmac::verify(&key, req.finish_payload.as_bytes(), &base64::decode(&req.mac)?)?;

    let payload: PairFinishPayload = serde_json::from_str(&req.finish_payload)?;
    finish_pair(&payload, None, u64::MAX, None)?;

    Ok(Json(()))
}

/*
 * Delegation
 */
async fn get_finish_payload(Path(challenge_id): Path<String>) -> AppResult<impl IntoResponse> {
    let payload = FINISH_PAYLOADS.get(&challenge_id)
        .map(|p| p.clone())
        .ok_or(HttpError::NotFound)?;
    Ok(payload)
}

async fn post_finish_payload(
    Path(id): Path<String>,
    payload: String,
) -> AppResult<impl IntoResponse> {
    require(PAIRING_CHALLENGES.contains_key(&id))?;
    require(!FINISH_PAYLOADS.contains_key(&id))?;

    // Raw string for flexibility
    debug!("Delegated pair finish payload: {:?}", payload);
    FINISH_PAYLOADS.insert(id, payload);
    Ok(Json(()))
}

async fn finish_delegated(
    Path(id): Path<String>,
    envelope: Json<SignedRequestEnvelope>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
) -> AppResult<impl IntoResponse> {
    let req: Delegation = envelope.open_for_delegation(&addr)?;
    debug!("Finish delegated pair: {:?}", req);

    // Prevent delegator from changing request
    {
        let orig_payload = FINISH_PAYLOADS.get(&id)
            .ok_or(HttpError::NotFound)?;
        require(*orig_payload == req.finish_payload)?;
    }

    let payload: PairFinishPayload = serde_json::from_str(&req.finish_payload)?;
    finish_pair(
        &payload,
        Some(envelope.device_id.clone()),
        req.expires_at,
        req.allowed_entities,
    )?;

    Ok(Json(()))
}

/*
 * Common
 */
async fn get_challenge() -> impl IntoResponse {
    let challenge = PairingChallenge {
        id: generate_secret(),
        timestamp: now(),
        is_initial: !STORE.has_paired_devices(),
    };

    PAIRING_CHALLENGES.insert(challenge.id.clone(), challenge.clone());
    Json(challenge)
}

pub fn service() -> Router {
    Router::new()
        .route("/initial/start", post(start_initial))
        .route("/initial/finish", post(finish_initial))
        .route("/delegated/:challenge_id/finish_payload", get(get_finish_payload))
        .route("/delegated/:challenge_id/finish_payload", post(post_finish_payload))
        .route("/delegated/:challenge_id/finish", post(finish_delegated))
        .route("/get_challenge", post(get_challenge))
}
